### PR TITLE
[FIRRTL] Implement initial verifications for ConnectOp.

### PR DESCRIPTION
**Pull request by mikeurbach**
This is part of https://github.com/llvm/circt/issues/232. This checks
that the connected values have equivalent types, as defined by Section
4.5 of the FIRRTL spec.

### CODE REVIEW

File: test/FIRParser/basic.fir
Comment (mikeurbach) [inline test/FIRParser/basic.fir:82]: Since `_t_2` is connected to `_t`, they have to have the same number of elements. I'm assuming it was intended for them to be valid to connect, so I've updated `_t_2` to match the number of elements in `_t`.

File: test/FIRParser/basic.fir
Comment (mikeurbach) [inline test/FIRParser/basic.fir:181]: `auto` is unsigned, but the result of the expression `add(s8, SInt<8>("hAB"))` is signed, so I've added a new output parameter called `sauto` that is signed, and can be connected to here.

**Review by mikeurbach — COMMENTED** [review]
This is the meat of `ConnectOp` verifications: checking type equivalence. I'm working on the remaining TODOs, but I thought I'd checkpoint here and make a PR to see if you folks have any feedback.

One thing I've been thinking is we can probably do something slick with the helper functions that actually implement the logic. I think we could have them return `LogicalResult` instead of just `bool`, and in the failure cases, return a `Diagnostic`, which gets attached to the generic error about "type mismatch" to include exactly which branch of logic this failed on. I'll explore this a little bit.

Also, this logic found some invalid IR produced in the FIRParser test. I updated the test to make it valid, but please let me know if something else should be done there.

**Comment by mikeurbach** [comment]
Marking this a draft until I get some feedback on the linked issue.

File: test/FIRParser/basic.fir
Comment (lattner) [reply test/FIRParser/basic.fir:82]: Yeah I suspect that there are a lot of hand written tests that are buggy (because I don't have all the rules internalized).  Please feel free to adjust them as you encounter them. Thanks!

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/Ops.cpp:739]: Type equivalence is a generally defined FIRRTL thing, so I'd recommend splitting this out into a more general "areTypesEquivalent" predicate in FIRRTL/Types.cpp.

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/Ops.cpp:737]: I'd avoid referencing a section number (which can drift) and URL, you can just mention the FIRRTL spec by name.

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/Ops.cpp:740]: Is this about equivalence, or is this connect specific?  If it is connect specific, then I'd move it out of this routine.

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (lattner) [reply lib/Dialect/FIRRTL/Ops.cpp:739]: I'd start this function out with:

destType = destType.getPassiveType();
srcType = srcType.getPassiveType();

because flips can occur at many levels and (AFAIK) don't affect equivalence. If you want to be fancy, do that once in the non-recursive public entrypoint to the predicate.

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/Ops.cpp:766]: Please use a for loop instead of zip and all_of here.  The metaprogramming isn't necessary and forces tuples into the type system unnecessarily.

C++ is not a beautiful functional language :-)

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/Ops.cpp:785]: I think that FIRRTL supports connecting to inputs.  Am I wrong about this?  I am pretty sure this occurred in real code, even if not documented.  @seldridge do you know?

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/Ops.cpp:707]: Is this checking type equivalence or connect equivalence?  If the later, I'd suggest something like "areBundleElementEquivalent".  Per the comment below, please drop the tuple type, it is just obfuscating the code with `get<>`'s.

File: test/firrtl/connect.mlir
Comment (lattner) [inline test/firrtl/connect.mlir:3]: nice testcase!

**Review by lattner — COMMENTED** [review]
Thank you for tackling this!

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (seldridge) [reply lib/Dialect/FIRRTL/Ops.cpp:785]: Connecting to source flow objects is disallowed based on "flow" checking, but not based on type checking. For ground types this means you cannot assign an input.

E.g., the following circuit will fail during the `CheckFlows` pass.

```
# cat Foo.fir
circuit Foo:
  module Foo:
    input in: UInt<1>
    in <= UInt<1>(0)
# firrtl -i Foo.fir
Exception in thread "main" firrtl.passes.CheckFlows$WrongFlow: : [module Foo]  Expression in is used as a SinkFlow but can only be used as a SourceFlow.
```

Note that this isn't a blanket "you can't connect to inputs" just that source flow can't be a LHS in an expression. If you make the input a bundle and put flips you can assign to the correctly flowing portions. E.g., the following is fine:

```
# cat Foo.fir
circuit Foo:
  module Foo:
    input in: { flip a: UInt<1> }
    in.a <= UInt<1>(0)
```

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (mikeurbach) [reply lib/Dialect/FIRRTL/Ops.cpp:766]: :) understood... and, i already fixed a bug related to getting the wrong tuple element...

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (mikeurbach) [reply lib/Dialect/FIRRTL/Ops.cpp:740]: I guess this is connect specific. This isn't mentioned in the equivalence section of the spec at all.

File: lib/Dialect/FIRRTL/Ops.cpp
Comment (mikeurbach) [reply lib/Dialect/FIRRTL/Ops.cpp:739]: Yep, this is going to be updated as you say. See my discussion in the [issue](https://github.com/llvm/circt/issues/232#issuecomment-733115018) for the reasoning that led me to the same conclusion.

**Comment by mikeurbach** [comment]
@lattner I resolved a couple of your comments (the big ones--related to removing `zip` and `all_of`, and just using passive types), but I haven't finished. The bot requested your review, but please hold off till I finish cleaning this up.

**Comment by mikeurbach** [comment]
@lattner @seldridge this PR is ready for review. Please take a look when you get a chance.

File: include/circt/Dialect/FIRRTL/Types.h
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Types.h:75]: I would make this comment (and later comments) clear that this is weaker type equivalence than is defined in the FIRRTL spec (namely that this is type equivalence of the passive types).

The spec explicitly calls out that `{a: {flip b: UInt}} is not equal to {flip a: {b: UInt}}`. However, our canonicalization makes these types of statements equivalent (and I'd argue they _are_ equivalent). I'm annoyed that a corollary of this statement is "canonicalization is disallowed".

In discussing this with @mikeurbach (as well as offline with @jackkoenig), I am fine with deviating from the spec here so long as we accept circuits which the Scala FIRRTL Compiler (SFC) rejects, but not if we reject circuits that the SFC accepts. (This assertion holds with this PR.)

File: test/firrtl/connect.mlir
Comment (seldridge) [reply test/firrtl/connect.mlir:3]: I agree! Thanks for painstakingly enumerating all these conditions. :+1:

**Review by seldridge — APPROVED** [review]
Looks great. Thanks for tackling this.

This PR is exposing some places where we deviate from the spec. As a rule I think that's fine so long as it only expands the set of FIRRTL circuits we accept and we're clear on how we deviate and why.

File: include/circt/Dialect/FIRRTL/Types.h
Comment (lattner) [reply include/circt/Dialect/FIRRTL/Types.h:75]: > In discussing this with @mikeurbach (as well as offline with @jackkoenig), I am fine with deviating from the spec here so long as we accept circuits which the Scala FIRRTL Compiler (SFC) rejects, but not if we reject circuits that the SFC accepts. (This assertion holds with this PR.)

This makes sense to me.

**Review by lattner — APPROVED** [review]
The code looks good to me, but I defer to @seldridge to review the logic.  Thank you for implementing this!!

File: include/circt/Dialect/FIRRTL/Types.h
Comment (mikeurbach) [reply include/circt/Dialect/FIRRTL/Types.h]: Sounds good to me. I updated the comment here and on the specific helper predicate for bundle element equivalence.

**Pull request merged**
